index.js

```diff
@@ -1,143 +1,165 @@
-var fs = require('fs');
-var http = require('http');
-var https = require('https');
-var _ = require('lodash');
-
-function NuxtServer(Nuxt, Builder, nuxtConfig) {
-    this._nuxt = Nuxt;
-    this._builder = Builder;
-    this._config = _.merge(nuxtConfig, {
-        cert: {
-            mode: 'http',
-            setup: {
-                http: {
-                    port: 8080
-                },
-                https: {
-                    port: 8443,
-                    path: {
-                        key: null,
-                        cert: null,
-                        ca: []
+const fs = require('fs');
+const http = require('http');
+const https = require('https');
+const consola = require('consola')
+const _ = require('lodash');
+
+const logger = consola.withScope('nuxt-server')
+
+module.exports = class NuxtServer {
+    constructor(Nuxt, Builder, nuxtConfig) {
+        this._nuxt = Nuxt;
+        this._builder = Builder;
+        this._config = _.merge(nuxtConfig, {
+            cert: {
+                mode: 'http',
+                setup: {
+                    http: {
+                        port: 8080
                     },
-                    handshakeTimeout: 120,
-                    requestCert: false,
-                    rejectUnauthorized: true,
+                    https: {
+                        port: 8443,
+                        path: {
+                            key: null,
+                            cert: null,
+                            ca: []
+                        },
+                        handshakeTimeout: 120,
+                        requestCert: false,
+                        rejectUnauthorized: true,
+                    }
                 }
             }
+        });
+
+        this._nuxt = new Nuxt(this._config);
+    }
+
+    addServerMiddleware(m) {
+        // Resolve
+        const $m = m
+        if (typeof m === 'string') {
+            m = require(m)
         }
-    });
-}
 
-NuxtServer.prototype.run = function (callback) {
-    var self = this;
-    function result(cb) {
+        if (typeof m.handler === 'string') {
+            m.handler = require(m.handler)
+        }
+
+        const handler = m.handler || m
+        const path = (
+            (m.prefix ? m.prefix : '') +
+            (typeof m.path === 'string' ? m.path : '')
+        ).replace(/\/\//g, '/')
+
+        handler.$m = $m
+
+        // Use middleware
+        if (_.isArray(handler)) {
+            handler.forEach((h) => {
+                this._nuxt.render.use(path, h)
+                logger.info('Added middleware to ' + path);
+            });
+        } else {
+            this._nuxt.render.use(path, handler)
+            logger.info('Added middleware to ' + path);
+        }
+    }
+
+    async run() {
```

Comment:
Some users of this module may be using callback instead of promises. In order not to break your applications, please keep the support.

Suggestion:

async run()  for async _runnable()

run(callback) {
    if (!callback) return this._runnable();
    const result = this._runnable();
    result.then(data => callback(null, data));
    result.catch(callback);
}